src/main/java/io/penguinstats/controller/v2/api/ResultController.java

```diff
@@ -0,0 +1,123 @@
+package io.penguinstats.controller.v2.api;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import io.penguinstats.constant.Constant.LastUpdateMapKeyName;
+import io.penguinstats.controller.v2.response.MatrixResult;
+import io.penguinstats.controller.v2.response.StageTrend;
+import io.penguinstats.controller.v2.response.TrendDetail;
+import io.penguinstats.controller.v2.response.TrendResult;
+import io.penguinstats.enums.Server;
+import io.penguinstats.model.DropMatrixElement;
+import io.penguinstats.service.DropInfoService;
+import io.penguinstats.service.ItemDropService;
+import io.penguinstats.util.CookieUtil;
+import io.penguinstats.util.LastUpdateTimeUtil;
+import io.swagger.annotations.ApiOperation;
+
+@RestController("resultController_v2")
+@RequestMapping("/api/v2/result")
+public class ResultController {
+
+	private static Logger logger = LogManager.getLogger(ResultController.class);
+
+	@Autowired
+	private ItemDropService itemDropService;
+	@Autowired
+	private DropInfoService dropInfoService;
+	@Autowired
+	private CookieUtil cookieUtil;
+
+	@ApiOperation("Get matrix")
+	@GetMapping(path = "/matrix", produces = "application/json;charset=UTF-8")
+	public ResponseEntity<MatrixResult> getMatrix(HttpServletRequest request,
+			@RequestParam(name = "is_personal", required = false, defaultValue = "false") boolean isPersonal,
+			@RequestParam(name = "show_closed_zones", required = false, defaultValue = "false") boolean showClosedZones,
+			@RequestParam(name = "server", required = false, defaultValue = "CN") Server server) {
+		logger.info("GET /matrix");
+		try {
+			String userID = isPersonal ? cookieUtil.readUserIDFromCookie(request) : null;
+			List<DropMatrixElement> elements = itemDropService.generateGlobalDropMatrixElements(server, userID);
+
+			if (!showClosedZones) {
+				Set<String> openingStages =
+						showClosedZones ? null : dropInfoService.getOpeningStages(server, System.currentTimeMillis());
```

Comment:
Arkplanner will have to query with showClosedZones=true

Reply:
I mean that in this if statement block, it may be always false, so maybe you can directly use `Set<String> openingStages = dropInfoService.getOpeningStages(server, System.currentTimeMillis());`